ТЗ11

nev-ivan opened the pull request.

Commit:
Первая итерация ТЗ11

Commit:
Первая итерация ТЗ11

Commit:
Первая итерация ТЗ11

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,101 +1,73 @@
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    public Map<String, String> filmNotFoundExceptionHandler(NotFoundException e) {
+        return Map.of("Ошибка", e.getMessage());
```
Стоит реализовать отдельный хендлер для исключений `@RestControllerAdvice`

Плюс, из методов обработчиков лучше будет возвращать более детализированные сообщения об ошибках в составе дтошки, например `ErrorResponse`

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,101 +1,73 @@
-    private final Map<Integer, Film> films = new HashMap<>();
-    private static final int MAX_DESCRIPTION_SIZE = 200;
-    private static final LocalDate EARLY_DATE = LocalDate.parse("1895-12-28");
+    private final FilmStorage filmStorage;
```
В контроллере стоит обращаться через сервисы, у нас выстроена цепочка controller->service->storage и ее стоит придерживаться

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -1,97 +1,84 @@
+    public UserController(UserService userService, UserStorage userStorage) {
+        this.userService = userService;
+        this.userStorage = userStorage;
+    }
```
Явный конструктор можно заменить на аннотацию lombok `@RequiredArgsConstructor`

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -4,19 +4,26 @@
         this.duration = duration;
     }
 
+    @Override
+    public int compareTo(Film other) {
+        return Integer.compare(other.getLikes().size(), this.getLikes().size());
+    }
```
Логику по сортировке фильмов по лайкам в обратном порядке стоит реализовать в самом методе сервиса, который как раз и отдает эту выборку фильмов по лайкам. Что если нам понадобится другая сортировка фильмов, например по продолжительности и тд. Через метод compareTo() это уже не реализуемо

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -4,19 +4,26 @@
+        return Integer.compare(other.getLikes().size(), this.getLikes().size());
+    }
 
+    private Set<Integer> likes = new HashSet<>();
```
Поля в классе объявлять лучше по значимости, от айди до менее значимых

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,74 @@
+    private final FilmStorage filmStorage;
+    private final UserStorage userStorage;
+
+
+
```
Лишние отступы

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    public List<Film> popularFilms(long count) {
+        if (count <= 0) {
+            throw new ValidateException("Неверный параметр count");
+        }
```
Можно провалидировать аннотацией на уровне контроллера, статься про валидацию [struchkov.dev/blog/ru/spring-boot-validation/](https://struchkov.dev/blog/ru/spring-boot-validation/)

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,74 @@
+        if (count <= 0) {
+            throw new ValidateException("Неверный параметр count");
+        }
+        if (filmStorage.findAll().isEmpty()) {
+            return null;
+        }
```
А зачем явно это проверять и возвращать null, stream() не выполнится если список пустой и вернется пустой список\

Возвращать null явно плохая практика

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,74 @@
+                .toList();
+    }
+
+    private void checkId(int id) {
+        List<Integer> allId = filmStorage.findAll().stream()
+                .map(Film::getId)
+                .toList();
+        if (!allId.contains(id)) {
+            throw new NotFoundException("Неизвестный фильм");
+        }
+    }
+
+    private void checkUser(int userId) {
+        List<Integer> allUserId = userStorage.findAll().stream()
+                .map(User::getId)
+                .toList();
+        if (!allUserId.contains(userId)) {
+            throw new NotFoundException("Неизвестный пользователь");
+        }
+    }
```
На уровне стораджа есть хешмапы, благодаря которым можно сделать быструю проверку на `containsKey()`, с такими проверками теряется все их преимущество

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,74 @@
+        checkId(filmId);
+        checkUser(userId);
+        Film film = filmStorage.getFilm(filmId);
+        Set<Integer> likes = film.getLikes();
+        likes.remove(userId);
+        film.setLikes(likes);
+        filmStorage.update(film);
```
Много лишних строк кода, достаточно запросить фильм и вызвать удаление, мы работаем с ссылкой на исходный объект в памяти. `filmStorage.getFilm(filmId).remove(userId);`

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
К этому сервису так же относятся комментарии из сервиса фильмов, логика тут очень похожая

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    public List<Film> findAll() {
+        return films.values().stream().toList();
```
Можно не через стримы, а через обычный конструктор листа 
`return new ArrayList<>(films.values());`

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -0,0 +1,96 @@
+        }
+
+        if (film.getId() == null) {
+            log.warn("ConditionsNotMetException");
```
В логах лучше писать более подробно, чтобы было проще отлаживать

Commit:
Вторая итерация ТЗ11

Commit:
Вторая итерация ТЗ11

Commit:
Вторая итерация ТЗ11

Commit:
Вторая итерация ТЗ11

Commit:
Вторая итерация ТЗ11

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/exception/ConditionsNotMetException.java:
```diff
@@ -1,5 +1,9 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
```
Не думаю, что в этом случае нужно возвращать 500 ошибку, обычно этот статус возвращается при ошибках на стороне сервера (недоступность сервера, таймауты при ответе и тд)

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -19,56 +16,35 @@ public class FilmService {
-        filmStorage.update(film);
+        filmStorage.checkFilm(filmId);
+        userStorage.checkUser(userId);
+        filmStorage.getFilm(filmId).getLikes().add(userId);
```
Лучше инкапсулировать логику добавления/удаления лайков и друзей в классах `User` и `Film`. Сейчас коллекции модифицируются напрямую, что может привести к багам, если в будущем будет добавлена дополнительная логика (например, проверка дубликатов и т.п.)

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -4,58 +4,51 @@
+        userStorage.checkUser(user1id);
+        userStorage.checkUser(user2id);
         User user1 = userStorage.getUser(user1id);
         User user2 = userStorage.getUser(user2id);
```
Зачем дважды обращаться к стораджу, если нам нужна сущность изначально, можно тогда в методе `getUser()` и `getFilm()` реализовать через `Optional`, он помогает легче работать с nullable объектами

`checkFilm()` стоит использовать только в тех случаях, когда нам просто нужно проверить наличие пользователя без его получения, но даже эта потребность может быть закрыта реализацией методов `getUser()` и `getFilm()` через Optional

Commit:
Третья итерация ТЗ11
- полностью удалены методы checkUser и checkFilm для проверки на null
- методы getUser и getFilm реализованы через Optional
- логика по работе с друзьями и лайками реализована в моделях

Commit:
Третья итерация ТЗ11
- полностью удалены методы checkUser и checkFilm для проверки на null
- методы getUser и getFilm реализованы через Optional
- логика по работе с друзьями и лайками реализована в моделях
